Objects/object.c

```diff
@@ -1426,9 +1426,18 @@ _PyObject_GenericSetAttrWithDict(PyObject *obj, PyObject *name,
             res = PyDict_SetItem(dict, name, value);
         Py_DECREF(dict);
     }
-    if (res < 0 && PyErr_ExceptionMatches(PyExc_KeyError))
-        PyErr_SetObject(PyExc_AttributeError, name);
-
+    if (res < 0 && PyErr_ExceptionMatches(PyExc_KeyError)) {
+        if (PyType_IsSubtype(tp, &PyType_Type)) {
```

Comment:
I think this might be unsafe: the `PyDict_SetItem` on line 1426 could trigger a destructor if there's already an object in the dict, and that destructor can run arbitrary Python code, so it could set `obj.__class__` to something else and cause `tp` to be garbage collected. To be safe, you should INCREF tp around line 1384 and DECREF it again in the done block.

Reply:
Thanks, updated.